[AArch64] Optimize memset to use NEON DUP instruction for more sizes

osamakader opened the pull request:
This change improves memset code generation for non-zero values on AArch64 by using NEON's DUP instruction instead of
the less efficient multiplication with 0x01010101 pattern.

For small sizes, the value is extracted from a larger DUP.  For non-power-of-two sizes, overlapping stores are used in some cases.

TargetLowering::findOptimalMemOpLowering is modified to allow explicitly specifying the size of the constant in cases where the constant is larger than the store operations.

Fixes #165949

osamakader committed:
Optimize AArch64 memset to use NEON DUP instruction for small sizes

This change improves memset code generation for non-zero values on AArch64
for sizes 4, 8, and 16 bytes by using NEON's DUP instruction instead of
the less efficient multiplication with 0x01010101 pattern.

Changes:
1. In SelectionDAG.cpp: For AArch64 targets, generate vector splats for
   scalar i32/i64 memset operations, which are then efficiently lowered to
   DUP instructions.

2. In AArch64ISelLowering.cpp: Modify getOptimalMemOpType and
   getOptimalMemOpLLT to return v16i8 for non-zero memset operations of
   any size when NEON is available (previously only for sizes >= 32 bytes).

3. Update test expectations to verify the new DUP-based code generation
   for both NEON and GPR code paths.

The optimization is restricted to AArch64 only to avoid breaking RISCV
and X86 tests.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8543,6 +8543,20 @@ static SDValue getMemsetValue(SDValue Value, EVT VT, SelectionDAG &DAG,
+  const TargetMachine &TM = DAG.getTarget();
+  if (NumBits > 8 && VT.isInteger() && !VT.isVector() &&
+      (NumBits == 32 || NumBits == 64) &&
+      TM.getTargetTriple().getArch() == Triple::aarch64) {
```
We can't have if Arch specific checks in the generic SelectionDAG code.

davemgreen commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -29702,6 +29708,31 @@ AArch64TargetLowering::EmitKCFICheck(MachineBasicBlock &MBB,
+  // a smaller scalar (e.g., i32 for a 4-byte memset).
+  if (FixedVectorType *VTy = dyn_cast<FixedVectorType>(VectorTy)) {
+    // Only handle v16i8 splat (128 bits total, 16 elements of 8 bits each)
+    if (VTy->getNumElements() == 16 && VTy->getElementType()->isIntegerTy(8)) {
```
Is it worth adding v8i8, or does it not get produced?

davemgreen reviewed (commented):
Can you add some extra test cases for constants that are not 0? I think they should be OK if they are i8 but it is worth having tests. It looks like there are a number of others that could do with updating too.

osamakader committed:
AArch64: move memset optimize to AArch64 Lowering

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
[AArch64] Optimize memset to use NEON dup for non-power-of-two sizes

This commit extends the NEON DUP optimization to handle non-power-of-two
memset sizes (e.g., 5, 6, 7, 9, 10, 12, 14, 15 bytes) by decomposing them
into multiple vector-extracted stores, matching GCC's behavior.

Key changes:
1. Extended shallExtractConstSplatVectorElementToStore to support 8-bit,
   16-bit, 32-bit, and 64-bit extractions from v16i8 splats.
2. Updated findOptimalMemOpLowering override to decompose non-power-of-two
   sizes into the largest possible stores (i64, then i32, then i16, then i8).
3. Ensured v16i8 remains the LargestVT so getMemsetValue generates the
   efficient dup v0.16b instruction.

Before: Small non-power-of-two memsets used the MUL pattern:
  mov w8, #0x01010101
  mul w8, w9, w8
  str w8, [x0]
  strh w8, [x0, #8]

After: All sizes now use NEON DUP:
  dup v0.16b, w1
  str d0, [x0]
  str h0, [x0, #8]

This matches GCC's output and improves code quality for memset operations.
The optimization works for all non-zero memset sizes, with only the final
i8 byte store falling back to GPR (due to LLVM instruction selection
limitations), which is acceptable.

Test updates reflect the new optimized output for non-power-of-two sizes

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

efriedma-quic commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8961,6 +8958,12 @@ static SDValue getMemsetStores(SelectionDAG &DAG, const SDLoc &dl,
   for (unsigned i = 0; i < NumMemOps; i++) {
     EVT VT = MemOps[i];
     unsigned VTSize = VT.getSizeInBits() / 8;
+    // Skip stores when Size is already 0. This can happen when an oversized
```
I'm not sure I understand how you end up in this situation.

The target specifies the type of every store.  That means the total size of all the types specified in MemOps should be the total size of the memset plus an adjustment for overlapping stores.  If the target is specifying too many types in the list, that seems like a bug in the target code.

osamakader committed:
[AArch64] Add assertions for findOptimalMemOpLowering correctness

Address reviewer feedback by adding proper assertions to catch bugs in
target implementations of findOptimalMemOpLowering.

Changes:
1. Add assertion that Size > 0 for stores that will be emitted (non-oversized)
2. Skip emitting oversized stores when Size == 0 (they were only added
   to MemOps for LargestVT selection to generate vector splats)
3. Add assertion that Size == 0 after processing all stores
4. Fix Size calculation for oversized overlapping stores to only subtract
   remaining bytes, not the full store size
5. Update comments to clarify the design

This ensures that targets don't add store types to MemOps that exceed the
required size, and violations are caught as bugs. The AArch64 implementation
correctly adds v16i8 as an oversized store only for LargestVT selection
when decomposing non-power-of-two memset sizes, and it's properly skipped
when Size == 0.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

efriedma-quic commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -8961,6 +8958,20 @@ static SDValue getMemsetStores(SelectionDAG &DAG, const SDLoc &dl,
+    // (with the last store potentially being oversized for overlapping stores).
+    // If Size is 0, all bytes have been written by previous stores. The only
+    // valid case for Size == 0 is when the last store in MemOps is oversized
+    // and was added only to ensure LargestVT selection for value generation.
```
Oh, I see.

I think I'd prefer to make findOptimalMemOpLowering return the LargestVT as a separate variable, instead of sticking it at the end of the array.  (It shouldn't be too hard to adjust the findOptimalMemOpLowering API; it isn't used in very many places.)

osamakader committed:
[AArch64] Refactor findOptimalMemOpLowering to use separate LargestVT parameter

Address reviewer feedback by refactoring findOptimalMemOpLowering to return
LargestVT as a separate output parameter instead of adding it to MemOps.

Changes:
1. Add optional EVT *LargestVT parameter to findOptimalMemOpLowering API
2. AArch64 implementation sets *LargestVT = v16i8 instead of adding it to MemOps
3. SelectionDAG.cpp uses LargestVT if set, otherwise computes from MemOps
4. Remove redundant shallExtractConstSplatVectorElementToStore calls since
   we already know VectorTy is v16i8 in that context
5. Update all call sites
6. Update SystemZ override to pass through the parameter

This avoids adding store types to MemOps that aren't actually emitted, which
was the reviewer's concern. The LargestVT parameter is only used for value
generation (e.g., to generate the vector splat for dup v0.16b), not for
determining which stores to emit.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

efriedma-quic reviewed (approved):
LGTM

efriedma-quic commented:
LLVM normally squashes commits, so your commit messages on individual commits won't make it through; it just uses the title of the PR and the contents of the initial comment in the PR thread.  I tried to update the title and summary, but please take a look and make any additional modifications if you want.

Also, the bot isn't triggering here, but:

> ⚠️ We detected that you are using a GitHub private e-mail address to contribute to the repo.
Please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account.
See [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#email-addresses) and [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.
